### ISSUE TITLE
Define a standard for custom URL loaders

### ISSUE THREAD

**Issue by nex3** [sass/sass#2535]
Many CSS features involve referring to assets outside the stylesheet using URLs: plain CSS `@import`s, background images, and so on. It's not always obvious how these assets will map to URLs on the server, though. Currently Sass provides no built-in support for this mapping, but doing so would be very useful.

I think Sass implementations should expose the ability for users to define their own URL-resolution logic that's invoked for every `url()` and every plain-CSS `@import`. This should be done through a JS API that's shared between Dart Sass and Node Sass, as well as a Dart API for Dart Sass users who invoke it via Dart.

I also think it would be valuable to provide some built-in resolution strategies that are available via the command line and through JS/Dart as standardized resolvers. For example, we might want to provide the ability to resolve URLs to `data:` URLs that embed the document directly, or to map directories on the filesystem to URLs that will be served.

- [ ] https://github.com/sass/dart-sass/issues/195
- [ ] https://github.com/sass/libsass/issues/532

**Comment by dwt** [comment]
As was discussed in #1015 I'd like to get some design considerations going here.

I know you are talking about custom URL-Resolvers for this issue, but I think there should be some serious thinking about a either a much simpler approach, or about exposing them through a much simpler interface to solve a very big subset of the use cases this is required for.

My thinking is that there are two modes of operating on included URLs - either you leave them alone (AFAIK the current approach) or you touch them. If you touch them, one very simple and very effective way, is to allow relative URLs, that are interpreted as relative to the *current* file (just as in CSS). When compiling these URLs, they should retain the this relative quality, but still modified to point to the original file.

As was mentioned this has problems when compiling sass to stdout, in which case the compiler doesn't know how to change those URLs.  Some options that are possible here:

* Warn and leave them as is, the user is required to put the referenced files in the same relative spot
* Error and abort compilation unless the user provides the information where the code will end up relative to the assets
* Allow in-file configuration by a sass directive that describes where the code will end up if compiled to stdout (maybe support two directives, so you can set a default, or override the url transformation)

Supporting custom loaders seems to require much more configuration - so maybe a simplified approach like this can be considered - especially because (I think) that it can be a really nice increment point to get to the full url loader feature, as it can be introduced by adding a sass function `file-relative-url()` (or whatever you call it that marks the urls to be specially treated). Then the URL-resolver feature you are discussing here can then later add different functions or marker arguments to this function to choose different resolver strategies.

**Comment by jhnns** [comment]
I've seen that issue popping up in a lot of webpack projects which usually take a more modular approach to importing and compiling Sass.

[Less](http://lesscss.org/) supports a `relativeUrls` option for several years now and I think it's a good and pragmatic approach to that problem (at least, it worked pretty well in my projects). The option has been [recently renamed to `rewriteUrls`](https://github.com/less/less.js/pull/3248) for better interoperability with CSS modules.

The idea is:

- `rewriteUrls: "off"` = do not touch `url()`
- `rewriteUrls: "all"` = rewrite all `url()` in imported files so that they can be resolved correctly from the main file
- `rewriteUrls: "local"` = only rewrite `url()` that start with a `.`. CSS modules have resolving semantics similar to Node.js.

[Further information](https://github.com/less/less-docs/pull/491/files)

There are two things to consider:

1. `url()` with expressions in the beginning cannot be rewritten, which is fine. The expression was probably there to work around that problem in the first place:

```scss
url("#{$asset-path}/image.png"); // can not be rewritten
```

Bootstrap used to have an `$asset-path` or something similar if I recall it correctly.

It would be possible to rewrite *that*, though:

```scss
url("./folder/#{$some-var}");
// could be rewritten to
url("./even/deeper/folder/#{$some-var}");
```

2. If the CSS code is supposed to be printed to stdout, the base path is unknown. In that case it would be necessary to configure a base directory or something like that which could default to cwd.

---

A different, also rather pragmatic approach would be to introduce file local variables like Node's `__filename` and `__dirname`. However, in order to avoid absolute file paths in the final CSS, they also would need to be relative from the output directory.

---

This problem needs to be solved somehow because right now it's impossible to use `url()` in an imported file or partial without knowing what the main file will be. It makes sharing Sass code via NPM really hard.

**Comment by nex3** [comment]
I think there are two orthogonal questions here:

### How does Sass know which file a URL is referring to?

This question is particularly tricky for a number of reasons:

1. Existing `url()`s in existing Sass files are presumably already well-formed URLs that are valid in the context of a browser resolving files.
2. Sass supports importers that may load stylesheets from places that aren't on disk at all, and those stylesheets may need to refer to assets.
3. A stylesheet may need to refer to an asset that's loaded by another importer (for example, an image from a library that's loaded via load path).

This is a domain where having global configuration like the custom loader I proposed originally isn't a great fit. Individual packages need to be able to refer to (at least) their own assets in a way that works regardless of how the downstream consumer configures their compilation. @dwt's `file-relative-url()` function is a pretty explicit way of declaring where an asset comes from that doesn't risk stepping on the toes of existing stylesheets. It doesn't cleanly address issues 2 or 3, though.

Another possibility would be to have an `asset-url()` function that does essentially an `@import`-style lookup of its URL by testing for a relative path, then passing it to each importer in turn. However, figuring out what those importers *return* is a tricky question.

Less's solution is to just rewrite some or all `url()`s, which I don't like very much. It depends on both global configuration, so that individual packages can't ensure that their URLs will be rewritten, and heuristics, so that two different `url()`s that look superficially similar may not behave the same way at all. That's something we try hard to avoid in Sass.

### How does the browser find the files?

Assuming the previous question is answered satisfactorily and we know which file on disk every `url()` (or `file-relative-url()` or `asset-url()`) corresponds to, we need to convert them to `url()`s that the browser can understand. How this conversion happens should ultimately be up to the application that's compiling the stylesheet, which means it probably is a good use case for a plugin (not necessarily a URL *loader*, but perhaps a *rewriter*). We could also build in some reasonable default rewriters, like "relative URL from the CSS file to the asset" or "`data:` URL that inlines the contents of the asset".

**Comment by jamesarosen** [comment]
This could be, but needn't be the same mechanism as for resolving Sass/SCSS paths on disk. Projects like [node-sass-tilde-importer](https://github.com/matthewdavidson/node-sass-tilde-importer) have begun to establish the standard that `@import "~@foo/bar";` means more or less `@import "#{path.resolve("@foo/bar/index.scss", "the project's node_modules/")}`. Unfortunately, that only works with node-sass.

**Comment by glen-84** [comment]
I'm using the sass-loader in webpack, and had to include resolve-url-loader to fix the relative URL weirdness, but since it relies on source maps, it's not possible to disable them (which you sometimes want to do, since they take a long time to generate).

Maybe if Sass offered something like:

```js
urlRewriter: (url, currentPath, done)
```

... we could write something to rewrite the URLs relative to the current path, or output folder, or whatever it needs to be.

**Comment by jslegers** [comment]
I don't really need a custom URL loader for the type of projects I work on per se, and it seems needlessly complicated to me. However, I do need a way to use URLs that are relative to file they are defined in, which I expected to be the default behavior with SCSS `@import` statements, as it is with CSS `@import` statements.

The type of projects I work on, are typically frameworks, libraries & tech demos / samples. Reusability is pretty critical for these types of projects, and both absolute paths or parts relative to the file that contains the CSS output are unacceptable, because the SCSS code I write is intended for use in many different projects, with unpredictable architectures.

If I create for example a pan control or zoom control, I do not know in advance how the customer will use the SCSS that goes with that control. Will they import them in React components? Will they import them in an Angular project? Will they use Webpack? Rollup? Svelte? Or maybe they just use Node-sass to produce a single CSS output & Dojo to pack their JavaScript. The code I write is supposed to work in all of these contexts, along with our own sample code, and this is only possible when URLs are interpreted as relative to the file they are found in, as is the case for native CSS `@import` statements. See #2701.

If it is not possible for paths in  `@import` ed SCSS files to behave the same as paths in  `@import`ed CSS files statements by default, I suppose a custom URL loader that allows me to define this behavior for a specific project would be an acceptable alternative, although I'd rather just do something like this :

````js
sass.render({
  /* The main SCSS file, importing all SCSS to be loaded */
  file: "app.scss",
  /* Rewrites relative paths, interpreting them as relative to the file they were found in */
  rewritePaths: true 
}, function(err, result) {
  /* ... */
});
````

Setting `rewritePaths` to `false` would maintain the current behavior of SCSS.

Setting `rewritePaths` to `true` would mean that all relative URLs would be interpreted as relative to the file they are found in, and thus rewritten as URLs relative to the path of the output file. 

Besides increasing the reusability of components written in SCSS, setting `rewritePaths` to `true` would also have the advantage that relative URLs behave the same as with native CSS `@import` statements, which means that you can freely switch between CSS imports & SCSS imports without worrying about URLs breaking.

> Less's solution is to just rewrite some or all url()s, which I don't like very much. It depends on both global configuration, so that individual packages can't ensure that their URLs will be rewritten

Why not allowing the behavior of URLs in `@import`ed files to be defined globally and also allowing it to be redefined locally?

For example, you could combine the aforementioned `rewritePaths` flag, with the following additions to the `@import` syntax :

````scss
@import norewrite("path/to/a/third/party/library");
@import rewrite("path/to/another/third/party/library");
````

The `norewrite` function would override the `rewritePaths` flag for the `path/to/a/third/party/library` and all of the files imported inside it, recursively, setting it to `true`.

The `rewrite` function would to the exact same thing, but setting the flag to `false`.

This way, you can import your old legacy code or third party dependencies and ensure the URL rewriting behavior is correct for those dependencies, regardless of the config of the project you're `@import`ing it to.

I'm not convinced this is very useful, though, as I can't think of any use cases where I would NOT want a URL to be rewritten to a URL relative to the output, considering the limited reusability of dependencies that do not rewrite URLs. IMO Less is correct to rewrite by default, and the current Sass behavior makes no sense whatsoever. Either way, this approach would avoid rewriting URLs for eg. old third party dependencies or your own  legacy code, where you may not want it, while still being able to globally set the default to rewriting URLs for everything else.

**Comment by nex3** [comment]
> I don't really need a custom URL loader for the type of projects I work on per se, and it seems needlessly complicated to me. However, I do need a way to use URLs that are relative to file they are defined in, which I expected to be the default behavior with SCSS `@import` statements, as it is with CSS `@import` statements.
> 
> The type of projects I work on, are typically frameworks, libraries & tech demos / samples. Reusability is pretty critical for these types of projects, and both absolute paths or parts relative to the file that contains the CSS output are unacceptable, because the SCSS code I write is intended for use in many different projects, with unpredictable architectures.
> 
> If I create for example a pan control or zoom control, I do not know in advance how the customer will use the SCSS that goes with that control. Will they import them in React components? Will they import them in an Angular project? Will they use Webpack? Rollup? Svelte? Or maybe they just use Node-sass to produce a single CSS output & Dojo to pack their JavaScript. The code I write is supposed to work in all of these contexts, along with our own sample code, and **this is only possible when URLs are interpreted as relative to the file they are found in**, as is the case for native CSS `@import` statements. See #2701.

*(Emphasis mine)*

The point I was trying to make in #2701 is that your sample code isn't guaranteed to work *even when* relative URLs are swizzled to be relative to the output file. For example, suppose your package contains `common/balloon/close-icon.png` and `common/balloon/luciad-balloon.scss` with `background-image: url(close-icon.png)`. A user then installs this via npm, so their package contains `node_modules/luciad/common/ballon/close-icon.png` *et al*. Then they import `luciad-balloon.scss` in a Sass file that gets compiled to `dist/css/index.css`. What should that `url()` produce?

If the URL is made relative to the output file, it will become `url(../../node_modules/luciad/common/balloon/close-icon.png)`. But their server is serving `dist/`, which means `node_modules` isn't being served at all! They may also be serving the icon from a CDN or something like that—as you say, there's no way to know what your users' downstream configuration looks like.

That's why a *configurable* solution is necessary. Ultimately, the end user (or their asset manager) needs to be able to fully control how these URLs are transformed. Relative URLs are not especially better as a default than any other solution. If we were to introduce a new default (which we probably won't because it would be a hugely breaking change) we'd probably have it be base64-encoded asset contents, since that's guaranteed to work anywhere.

> If it is not possible for paths in `@import` ed SCSS files to behave the same as paths in `@import`ed CSS files statements by default, I suppose a custom URL loader that allows me to define this behavior for a specific project would be an acceptable alternative, although I'd rather just do something like this :
> 
> ```js
> sass.render({
>   /* The main SCSS file, importing all SCSS to be loaded */
>   file: "app.scss",
>   /* Rewrites relative paths, interpreting them as relative to the file they were found in */
>   rewritePaths: true 
> }, function(err, result) {
>   /* ... */
> });
> ```
> 
> Setting `rewritePaths` to `false` would maintain the current behavior of SCSS.
> 
> Setting `rewritePaths` to `true` would mean that all relative URLs would be interpreted as relative to the file they are found in, and thus rewritten as URLs relative to the path of the output file.
> 
> Besides increasing the reusability of components written in SCSS, setting `rewritePaths` to `true` would also have the advantage that relative URLs behave the same as with native CSS `@import` statements, which means that you can freely switch between CSS imports & SCSS imports without worrying about URLs breaking.

As I said in the initial comment though, we'd want to provide some default resolution strategies. Swizzling relative URLs would certainly be one of these. This would probably look something like this:

```js
sass.render({
  file: "app.scss",
  urls: sass.url.relativeToCss,
}, function(err, result) {
  /* ... */
});
```

> > Less's solution is to just rewrite some or all url()s, which I don't like very much. It depends on both global configuration, so that individual packages can't ensure that their URLs will be rewritten
> 
> Why not allowing the behavior of URLs in `@import`ed files to be defined globally and also allowing it to be redefined locally?
> 
> For example, you could combine the aforementioned `rewritePaths` flag, with the following additions to the `@import` syntax :
> 
> ```scss
> @import norewrite("path/to/a/third/party/library");
> @import rewrite("path/to/another/third/party/library");
> ```
> 
> The `norewrite` function would override the `rewritePaths` flag for the `path/to/a/third/party/library` and all of the files imported inside it, recursively, setting it to `true`.
> 
> The `rewrite` function would to the exact same thing, but setting the flag to `false`.
> 
> This way, you can import your old legacy code or third party dependencies and ensure the URL rewriting behavior is correct for those dependencies, regardless of the config of the project you're `@import`ing it to.
> 
> I'm not convinced this is very useful, though, as I can't think of any use cases where I would NOT want a URL to be rewritten to a URL relative to the output, considering the limited reusability of dependencies that do not rewrite URLs. IMO Less is correct to rewrite by default, and the current Sass behavior makes no sense whatsoever. Either way, this approach would avoid rewriting URLs for eg. old third party dependencies or your own legacy code, where you may not want it, while still being able to globally set the default to rewriting URLs for everything else.

Generally speaking, it's not a great design to encode configuration (such as "how should URLs be emitted") into source files that may be shared across multiple contexts with multiple different needs.

**Comment by jslegers** [comment]
> For example, suppose your package `contains common/balloon/close-icon.png` and `common/balloon/luciad-balloon.scss` with `background-image: url(close-icon.png)`. A user then installs this via npm, so their package contains `node_modules/luciad/common/ballon/close-icon.png` et al. Then they import `luciad-balloon.scss` in a Sass file that gets compiled to` dist/css/index.css`. What should that `url()` produce?

By default, I would expect a relative link to the folder `node_modules/luciad/common/ballon/close-icon.png`, as it is done in Less.

If I want eg. a base64-encoded data-url or a url relative to eg. `dist/assets/balloon/close-icon.png` instead, I would not expect Sass to take care of that and I would look for a project like [`webpack-contrib/sass-loader`](https://github.com/webpack-contrib/sass-loader) to add additional transformation steps after Node sass or Dart sass converted my SCSS to intermediary CSS output. 

Converting SCSS to CSS & transforming CSS source code into an optimized production-ready custom format are two very different use cases. I don't think Sass is the right tool for the latter. Keeping relative URLs relative to the file that contains it, however, is a pretty basic SCSS to CSS tranformation step that is essential for creating reusable SCSS components, and therefore doesn't belong to the latter use case. See below.

> They may also be serving the icon from a CDN or something like that

Then we're talking about absolute URLs. URL rewriting is only relevant to relative URLs. 

At least it is with respect to SCSS to CSS compilation. Sure, you may want to rewrite absolute URLs at a later step, but, again, see below for more details on why I'm not convinced that's Sass's responsibility.

> there's no way to know what your users' downstream configuration looks like.

Sure... but common sense code conventions are the foundations of any solid software architecture.

As the SCSS source code level, relative paths & absolute paths have different use cases : 

* you should always use relative URLs when the relative "distance" between your resource & your SCSS source code is fixed but the position of those files within your folder structure is likely to change.

* you should always use absolute URLs when the relative "distance" between your resource & your SCSS source code is likelty to change but the position of the resources is fixed.

This is why URLs should ALWAYS be rewritten by default, and absolute URLs should NEVER be rewritten by default.

> That's why a *configurable* solution is necessary. Ultimately, the end user (or their asset manager) needs to be able to fully control how these URLs are transformed.

For situations where a resource folder is inaccessible be the web server for any reason, you will also need a way to easily copy any assets used from eg. `node_modules/path/to/asset` to `asset/img`. Because if you don't, no amount of URL rewriting is going to help you prevent additionally having to copy all resources from `node_modules/path/to/asset` to an `asset/img` either manually or through an external script, which is quite a hassle as well. Either that, or you'll need to embed a resources as a data-url. And, ideally, you'll want to define for each URL whether to embed or which folder to copy the resource to.

Therefore, I don't think you want Sass to determine the final URL for this type of projects. IMO this type of project should use an additional dependency, like Webpack, that is specialized in bundling all your resources together in an optimal way, allowing to you choose between embedding certain resources as data-urls while copying other images to an assets folder & rewriting the URL path for other resources in a highly configurable, custom way.

In a Webpack project, you'd typically use a dependencies like `webpack-contrib/sass-loader` & [`webpack-contrib/file-loader`](https://github.com/webpack-contrib/file-loader) where SCSS to CSS compilation by Dart or Node Sass is but one in number of small conversion steps between initial source code and the production code that will be served to the end customer.

I don't think it would be wise for Sass core to try to tackle such situation, as IMO this is way beyond the scope of a project like Sass. IMO, Sass shouldn't get involved in complex, configurable transpilation from source source to minified production code and it should leave this to specialized projects like `webpack-contrib/sass-loader`.

However, I don't believe that use case applies to keeping a URL relative to the file that contains it, as it is a very basic use case that is extremely common for even the most simple projects. Right now Sass is unable to use any urls but data-urls in a predictable, common sense, re-usable, zero-config, project-independent way. 

And, to makes matters worse, a URL in a CSS dependency will be interpreted differently depending on whether the dependency is embedded in my final output or loaded as an external dependency using the CSS `@import` statement. That makes this a major issue, IMO, and one that can not nor or should not be left to subsequent transformation steps. 

> Relative URLs are not especially better as a default than any other solution.

IMO, URLs relative to the path that contains the URL in your SCSS source code are a *much* better default than the current approach, because it is possible and easy to make those work EVERYWHERE without any configuration as long the path to your resource folder is accessible on your server.

URLs relative to the path that contains the URL in your SCSS source code allows you to create reusable components that are shared across projects and that can be put anywhere on your server without worrying about the URL breaking 

URLs relative to the path that contains the URL in your SCSS source code can be used eg. for dependencies you manually copy path in some `lib/component` or `asset` folder or dependencies that you install via npm, no matter where those folders are on your server, with the only conditions being that that path is available on the web server.

And for more complex use cases, you can always use a tool like Webpack to add additional transformation steps. See earlier.

> If we were to introduce a new default (which we probably won't because it would be a hugely breaking change)

IMO, you will gain far more than you'll lose when changing the default.

But yes, changing default always comes with a risk of messing up legacy code, which is why it makes sense to introduce a flag that allows you to switch the default behavior between old and new behavior.

And to make that transition still easier, you could set the flag to the old behavior for the first one or two new major releases while and change the flag at a later stage after which you deprecate the old behavior and possibly drop it one or two major releases later, giving devs time to gradually adopt a new behavior.

> ... we'd probably have it be base64-encoded asset contents, since that's guaranteed to work anywhere.

As I said previously in this comment, that is indeed the only solution that works for me.

However, as I also mentioned it's a poor solution, because it significantly increases the amount of data that needs to be loaded if you use the same image multiple times on multiple locations.

But sure, it would definitely be an improvement if Sass provided a native way to convert external resources to data-urls, whether or not it's a default.

Speaking of embedding as a data-url, this is the ONLY option I can think of that doesn't break my code in reusable modules. And I don't like that at all, considering data-urls are pretty hard to maintain and it can significantly increase the amount of data that needs to be loaded if I want to use the same image on multiple locations.

For small images, though, using a data-url is surely acceptable, although it would be a much better experience if there was a native function that converts any url to a data-url. I would expect a method like this (found [here](https://performancejs.com/post/ccffb43/Automatically-Generating-Base64-Inline-Images-With-SASS)) to be available in sass to alleviate the burden of maintaining base64-encoded data-urls.

Another workaround to Sass's inability to deal with relative urls would be the ability to do something like this :

````scss 
$path-to-collapse-component : parentdir(currentpath());

.collapse-component button {
  background : url(#{$path-to-collapse-component}/arrow-down.jpg);
}

// ...

$basepath : "/web";
$asset-folder : $basepath  + "/assets";

bg {
  background : url(#{$asset-folder}/site/background.jpg);
}
````

If I could have a way to determine at compile time what's the current absolute path of a particular file, I could calculate use absolute paths without having to worry that my paths are going to break by calculating the correct absolute path at compile time.

> As I said in the initial comment though, we'd want to provide some default resolution strategies. Swizzling relative URLs would certainly be one of these. This would probably look something like this:
> 
> sass.render({
>   file: "app.scss",
>   urls: sass.url.relativeToCss,
> }, function(err, result) {
>   /* ... */
> });

I still think it makes more sense to have that as a default at some point in the future, but as long as this feature is available I'll be very, very happy 👍 

> Generally speaking, it's not a great design to encode configuration (such as "how should URLs be emitted") into source files that may be shared across multiple contexts with multiple different needs.

If you want a dependency to be shared across multiple contexts with multiple different needs, you want to avoid the location of your dependency within your project or the configuration of own project breaking the behavior of that dependency. 

And this is impossible with Sass right now (1) *in a zero-config way*, if your component consists of SCSS files that contains external resources (like images or fonts) that (2) *come shipped with the component* and (3) *are not part of the component's public API*, because this requires URLs relative to the SCSS source file that contains it, which is impossible with the way Sass currently behaves!

**Comment by nex3** [comment]
I met with some Sass team members to discuss this. We think it makes sense to define a new built-in function (which we're tentatively calling `sass-url()`) that takes a string URL that refers to a file *on disk* and returns an unquoted `url()` string that refers to that file *on the server*.

We think a new function is better than allowing every `url()` to be rewritten because many existing Sass files already contain `url()`s that refer to files on the server in heterogeneous ways, which means our URL rewriting logic would somehow need to avoid breaking those existing references while also establishing a new convention for referring to local files. This is likely to be impossible, and even if it's not, it'll definitely involve a lot of migration churn that we'd just as soon avoid.

**How does `sass-url()` know which file a URL is referring to?** We'll add a new optional method to the custom importer API that returns the location of a resource referred to by a given URL, in the same way that importer returns Sass file contents. For almost all importers, this location should be a `file:` URL referring to a physical file on disk. (Note that this means that this feature is blocked on #2509).

The built-in filesystem importer that's used to load the entrypoint and to load files from load paths will support this method, which will allow `sass-url()` to refer to files relative to the file that contains it *and* files from packages on the load path. If popular import plugins also implement this method, you'll be able to write `sass-url("~@foo/bar/asset.png")` as well.

**How does the browser find the files?**

The location of the file returned by an importer (usually in the form of an absolute `file:` URL) will be passed to a URL rewriter plugin along with the location of the CSS file being generated (if that's known). This plugin will be responsible for converting that URL to one that's resolvable by browsers. This gives the downstream application maximal control over how its asset URLs are loaded.

Sass will also provide a few built-in resolvers that should cover most common use-cases. One of these will load the contents of files from disk and inject them as `data:` URLs; this one will be enabled by default, since it's the most foolproof method of ensuring the file contents makes it to the browser.

**Comment by yellow1912** [comment]
Is there any progress on this? If I want to completely turn off sass url resolve is this the feature I should look forward to? If not is there any way to do that?

**Comment by nex3** [comment]
@yellow1912 I'm not sure what you mean by "completely turn off sass url resolve". Sass currently does no URL resolution at all; this proposal would add additional resolution behavior.

**Comment by DeMoorJasper** [comment]
What's the progress on this? It would be really nice to finally be able to fix some of the url bugs that happen in both Parcel and webpack when using sass.

Any of the proposed solutions would resolve this, both `rewriteUrls` like less does and the callback function idea.

Don't wanna be pushy and totally respect the timeline of this project as it's quite large and widely used, just wanna know of any progress as there's only really been theoretical discussion so far.

If there's any place I can help I'd be happy to do so

**Comment by kamilic** [comment]
Is there any progress on this?

**Comment by alexander-akait** [comment]
@nex3 It is possible to achieve this with some hacks? Apparently the discussion has died down, but the problem still exists

**Comment by nex3** [comment]
No, this is something that won't be possible until we go through a full design process. If anyone wants to start that work by writing up a design spec, please go ahead—that's the best way to make progress here.

**Comment by alexander-akait** [comment]
@nex3 Do you have ideas where we should start? Maybe somebody already started it?

**Comment by ghost** [comment]
Whats the status? It kinda sucks to not be able to use ANY SCSS NPM PACKAGES (font packages, images etc.) because anytime I import/use them the relative urls won't be rebased and go nowhere. I can't fix this with something like `postcss-import` and `postcss-url` because they don't know where the imported scss module came from, this can only be solved during scss compilation itself. Thank you so much to everyone whos investing their time in this!

**Comment by pmcvay** [comment]
After reading these threads it sounds like sass just passes through the url, but I have a scss file with
```url('fonts/icomoon.ttf')``` 
and compiling it results in 
```url('/styles/fonts/icomoon.ttf')```

Why is this happening? I would prefer that it leaves the url alone. Or if it adds the `styles` at least make it `styles/fonts/icomoon.ttf` so it stays a relative path. It doesn't seem like its working the way this thread says it should. As a side note, I am using `sassc`

**Comment by canonic-epicure** [comment]
Why not to start with something simple and obvious, instead of trying to solve the whole huge use case at once? Similar to how TC39 adds features to JS - one by one. Imo, the use case of maintaining the relative urls is such a simple and obvious feature. Can be solved with `url_rel()` to make the user intention explicit and keep backward compat. 

The [original issue](https://github.com/sass/sass/issues/1015) about this problem was filed in 2013 and since then, SASS team seems to be discussing the ways of solving it. But as usual, instead of solving the immediate problem people are suffering from 2013, the discussion is about much more general "url rewriting" thing.

**Comment by Roy-Orbison** [comment]
I don't know why @canonic-epicure's [comment](https://github.com/sass/sass/issues/2535#issuecomment-888997351) was marked as spam, it seems to be an honest assessment of the current status.

Indeed it is a complex issue to completely resolve, evident from the detailed discussion, but there has been a pressing need need for _some_ solution for years. The simple, common case of referencing files within a single folder hierarchy could have been solved already.

Take this basic tree:
```none
style/
├─ main.scss
├─ asset/
│  ├─ _x.scss
│  ├─ x.woff2
```

As has been said, one would like to be able to write `url(x.woff2)` in `asset/_x.scss`, include it in `main.scss`, and have that compile to `url(asset/x.woff2)`. Having a function like `file-relative-url()` might be useful but doesn't seem feasible as a general solution because that would require it to be used by all included and 3rd-party library code. Providing an _optional_ method of resolving relative paths automatically at compile time makes the most sense. This could also allow a path prefix parameter that correlates with an actual served path, or a CDN base URL, or any token that could be used a search target for further processing. Sass could just skip resolution of absolute, scheme-relative and root-relative URLs. This alone could resolve a **huge** proportion of these issues.

@nex3 Has clearly explained how this would not cover the common and important case of dealing with library code outside an app's style code hierarchy, however that would be resolvable by a developer making use of symlinks.

The example tree becomes something like this:
```none
style/
├─ main.scss
├─ asset/
│  ├─ _x.scss
│  ├─ x.woff2
├─ external-asset/ -> ../node_modules/path/to/asset/
│  ├─ _y.scss
│  ├─ img/
│  │  ├─ y.svg
```
Using the earlier example, the compile target dir is `dist/css/` and `dist/` is the serving root. Automatically resolving the symlinked path (without a prefix) would produce `url(external-asset/img/y.svg)` in the file `dist/css/main.css`, which would finally resolve as the URL path `/css/external-asset/img/y.svg` when `/css/main.css` is served. Now `dist/css/` does not directly serve files in `node_modules/`, but again this is solvable by the developer using symlinks, etc. and is outside the scope of this proposed solution.

Part of one's build process could be that every dir (and symlinked dir) in `style/` is also symlinked in `dist/css/`, e.g. `dist/css/external-asset/ -> ../../style/external-asset/`; all non-sass files below `style/` could be copied/linked; or those files could be archived along with the files below `dist/` if packaging. If files below `dist/` but outside of `dist/css/` must be referenced, then one can use root-relative URLs ignored by the resolver, or use yet more symlinks below `style/` that can be used as tokens for later replacement in Sass' output. There are many ways to overcome such obstacles, but none that satisfactorily address the relative path issue.

The point is that giving Sass the ability to resolve relative paths may not be the ultimate solution, but it'd be a very workable starting point.

**Comment by halodevcr** [comment]
Hi there, just facing the exact same issue. Does anyone have a workaround ? a hack something that may help while we wait for this feature to be implemented ?

**Comment by rdhelms** [comment]
The `sass-url()` proposal from @nex3 sounded really promising here. Are there limitations or other issues preventing that from being implemented?

**Comment by Roy-Orbison** [comment]
@rdhelms That still leaves library code in the lurch. There are many good improvement proposals here and elsewhere, but they don't really address the core issue from which this one arose, i.e. data loss of relative paths during compilation. Any extension or custom URL handling shifts the responsibility of this data loss from Sass – where it's an easily solved problem – onto every developer that uses it. I'm yet to read anything from a Sass developer that acknowledges this.

If we had only the [same compilation options as Less](https://github.com/sass/libsass/issues/532) then this would be solved problem out-of-the-box for most of us, and provide a post-processing target for those with more complex requirements.

**Comment by rdhelms** [comment]
@Roy-Orbison Thanks for that info! I think you're definitely right that a sass-url function wouldn't be perfect, although maybe it would still be better than nothing? I encountered this issue as I was working on an npm package used internally at our company, and since we have tight control of that package I would happily have updated the library to use a function like sass-url, and we would have been good to go.

For now, I ended up just moving the code using `url()` into a regular css file, since the relative paths work totally fine there 🤷🏼‍♂️

**Comment by Roy-Orbison** [comment]
@rdhelms Thank you for providing a textbook example: paths not resolved in Sass, could work around because you control the library code, but end up with a slightly hacky solution anyway!

**Comment by 75th** [comment]
> The point I was trying to make in https://github.com/sass/sass/issues/2701 is that your sample code isn't guaranteed to work even when relative URLs are swizzled to be relative to the output file. For example, suppose your package contains `common/balloon/close-icon.png` and `common/balloon/luciad-balloon.scss` with `background-image: url(close-icon.png)`. A user then installs this via npm, so their package contains `node_modules/luciad/common/ballon/close-icon.png` et al. Then they import `luciad-balloon.scss` in a Sass file that gets compiled to `dist/css/index.css`. What should that url() produce?

I think maybe a big part of the disconnect here is that this

> Then they import `luciad-balloon.scss` in a Sass file

and this

> that gets compiled to `dist/css/index.css

are two separate URL-rewriting issues, and what we desperately need is a simple solution to the first and not a comprehensive solution to the second.

When you `@import` or `@use` or `@forward` a Sass file in another Sass file, it doesn't seem to me like you need to know anything about the final compile destination to know how to manipulate the relative URLs you find in the imported file. You know what file you're operating on, and you know the relative path to the file you're including. At the point of `@use`, Sass knows everything it needs to know about how to manipulate those relative URLs, and nothing outside of Sass can possibly get this information after that point.

As to how to rewrite stuff when all the `@use`s have happened, the file is fully processed, and you're compiling to stdout or a file somewhere or whatever, *that* can be offloaded to other scripts or Webpack loaders or whatever. But `@use`d files can only be handled by Sass, and the final output format is irrelevant (or seems so to me) to those rewrites. 

(I could of course be wrong about any of the above and would appreciate ELI5 explanations of what I'm missing if I'm missing something.)

**Comment by nex3** [comment]
I don't think you can separate the question of "where are the URLs supposed to point" from "where is the CSS output" unless you want to completely disclaim support for relative URLs—which I don't. Relative URLs in compiled CSS output are resolved relative to the URL the CSS itself is served from, which is something the Sass compiler can't intrinsically know. It _must be_ provided by a plugin.

> You know what file you're operating on, and you know the relative path to the file you're including.

This is enough to produce an absolute `file:` URL for the asset in question, but an absolute `file:` URL isn't sufficient for actually serving the CSS to a user. 

> As to how to rewrite stuff when all the `@use`s have happened, the file is fully processed, and you're compiling to stdout or a file somewhere or whatever, _that_ can be offloaded to other scripts or Webpack loaders or whatever.

If we're going to go to the trouble of designing an entire URL loader plugin, it doesn't make sense to just say "...and it only emits absolute `file:` URLs that are useless until you post-process them again". We should give users the tools to create fully useful CSS output.

**Comment by Roy-Orbison** [comment]
>which is something the Sass compiler can't intrinsically know. It must be provided by a plugin.

I really disagree with this appraisal of the situation. It's like saying *unless we have a panacea that covers all edge cases, we can't do basic relative URL resolution.* Sass *could* provide sensible relative URL resolution without plugins and without resorting to file URLs. It's a minority of users that have assets **above** the entry point file in a directory tree, it's not how most Sass is structured. People typically write a main file (or files) as entry points, then have all partials and their associated assets adjacent to or below those in the directory tree. If they *do* have assets above a directory tree, Sass could still output `../` prefixes for those URLs, because at least that would be better than total loss of relative paths. If the URLs for assets outside the tree are broken now, they're not going to get more broken if a user can opt in to relative path resolution.

As [I outlined before](https://github.com/sass/sass/issues/2535#issuecomment-985181778), it is quite possible to restructure files, or use symlinks, etc. to produce a workable output from Sass compilation. There are enough options. What we *cannot* do, is get the relative path prefix that Sass followed to get to an included asset, post-compilation, as pointed out again by @75th:

>Sass knows everything it needs to know about how to manipulate those relative URLs, and nothing outside of Sass can possibly get this information after that point.

Also a `file:` URL would be a fantastic option because it's not a publishable scheme. Anyone who's capable of writing a custom URL loader plugin is very likely capable of writing something to replace `file:` URLs in a CSS file. I'd hazard a guess that the users who want a custom URL loader option would prefer `file:` URLs than nothing at all.

**Comment by nex3** [comment]
When we design a new feature, even if we decide to do a smaller version of the feature to make it easier to land, it's crucial that we do it in a way that won't stop us from doing the full version in the right way in the future. In some cases, this means it's easier to just specify the full version from the beginning to ensure that the simpler use-cases slot in nicely. I believe this is one of those cases, since any "plug-and-play" solution such as "emit paths relative to the CSS output" will need to be modeled internally and in the spec as a URL rewriter anyway.

If you're interested in seeing progress here, the best course of action would be to adopt https://github.com/sass/sass/pull/2927, address the outstanding comments, and clean it up to modern language/API proposal standards (the proposal as written is pretty sketchy and needs some fleshing out).

**Comment by MrFox131** [comment]
Sorry being rude, guys


But WHAT THE FUCK R U DOING HERE?

In my current project of fucking thousands of lines, we have a modular frontend. It is assembled separately for each module. And I fucking want to include scss files from external sources using a relative path. Yes, I can do that, but the url directive behaves like the last scum. Have you ever heard of the fact that the url directive can refer to local files that they want to include with various kinds of bundlers in their build? No? Then you are fucking deaf as wood grouse (Russian proverb). So, guys, you’ve been solving a problem for fucking 5 years that webpack and rollup solved for you, developers, because everyone fucking needs this feature. Can you even fucking imagine how much junk I will have in my project in the folder with static assets if I make a single directory for including them as a url directive for all frontend modules? But they fucking refer to each other. You should be fucking ashamed. Because everything you are discussing here is actually not worth a damn. But alas, I don’t have the time or energy to understand your fucking code and this fucking dart. But you fucking decided to develop something that is a de facto industry standard, so get your ass in your hands and save your users from the need to use crutches.

P.S. I wouldn’t bomb like this if you came to a conclusion at least a year in 2019 and at least left it as a ready-made solution to fit into the development of third-party developers and solve this pain. But the first mention of how much I knew during the fucking 2 weeks of researching the problem was in 2013. What the fuck have you been doing for 10 years? In 10 years, this is a junior developer who will examine everything, fix it, debug it and tell the seniors to fuck off, because they spent 10 years arguing and nothing more.

**Comment by Banner-Keith** [comment]
@MrFox131 People like you are the reason why a lot of open source maintainers give up on maintaining their own source projects or taking them closed source. You come in here and get all mad at these people for not solving this, but also make no effort to help. Do you think this comment is going to make them want to work on this more?

When you're using open source software, you are welcome to report an issue, but then you either use a workaround or open a PR if it isn't being worked on. Being rude to the maintainers is super unhelpful.

**Comment by MrFox131** [comment]
> @MrFox131 People like you are the reason why a lot of open source maintainers give up on maintaining their own source projects or taking them closed source. You come in here and get all mad at these people for not solving this, but also make no effort to help. Do you think this comment is going to make them want to work on this more?
> 
> When you're using open source software, you are welcome to report an issue, but then you either use a workaround or open a PR if it isn't being worked on. Being rude to the maintainers is super unhelpful.

I made really a lot of work for several open source projects. But I read ALL issues in a plenty of repos about this absolutely ridiculous issue. And in this repo several people said - it's easy to implement this feature in a vary basic way. If u know 5his project of course under the hood. And more over, maintainer above my comment(thx to him) noticed enough on the same topic.

I just really hate moments of empty useless talks when your users struggle. 

I think you can understand me too, just imagine the frustration that i get when it found out that for 10 years no progress was made. And all solutions are just workarounds inside another projects. 
I don't have my own opensource projects, but have several solo developed projects for free. For some non-profit organizations in my country

I just cant understand being such ignorant to the problem. Moreover, i see that even maintainer realizes how ridiculous it is

**Comment by MrFox131** [comment]
@Banner-Keith Maybe you're right, maybe it's TOO rude. I wrote it in the heat. I sincerely apologize and will probably delete that comment. But please start doing something. I'm writing a huge combine to solve this problem now in my code, although for Sass developers it's much less effort to implement internally and a huge number of people need it

**Comment by Banner-Keith** [comment]
@MrFox131 I'm not a maintainer of this project so no need to apologize to me. I've just seen a few great open source maintainers quit and I want to make sure we have a positive community here.

**Comment by septatrix** [comment]
This would also resolve #597 which was once closed when Sass was much younger but would be possible once this is implemented

**Comment by jneuhaus20** [comment]
Regarding

> Relative URLs in compiled CSS output are resolved relative to the URL the CSS itself is served from, which is something the Sass compiler can't intrinsically know.

and

> I really disagree with this appraisal of the situation. It's like saying unless we have a panacea that covers all edge cases, we can't do basic relative URL resolution. 

I disagree with @Roy-Orbison, bc what @nex3 said is absolutely true and sounds like it was meant to narrow the solution space to what the Sass compiler can reliably do.

1. The file will be resolved relative to where the output CSS is served.
2. The Sass compiler can make no guarantees and has no control beyond that.

Therefore the best you can do is make reasonable assumptions based on standards and conventions, and make it transparent how to coerce one's use case into something that works with it.

Fundamentally, you've got "where the resource is" and "where it will be".  _I_ know where it is at compile time, and _I_ (probably) know where it's going to go in the ultimate dist.  What I _don't_ know is what root all these tools are using to resolve, and what they're going to change it to.

I explicitly _don't_ want plugins to decide, because they all do it differently and make their own assumptions based on their very narrow scope. _I_ want to decide as the developer blessed with this responsibility, and be insulated from package updates breaking my build.

**Comment by jneuhaus20** [comment]
Case in point:

Angular 18 switched to esbuild. My files didn't move, and where they're served from didn't change. Specifically, 

1. The font files I'm serving with `uri()` are all in `assets/<path>` relative to the index.html file and the scss file at the top of the import chain.
2. This relative structure is exactly the same in my output folder, aside from `styles.scss` becoming `styles.css`.

This was the default structure, it makes sense to me, and I have full control over what in the `src\assets` folder is copied and to where.

But my tilde is an error now, and the paths that make the project build are getting rewritten to `/media/<filename>` (the whole relative path is flattened to one folder) and copied to a different place than before.   Also the tests still use karma and webpack, which do things differently, so anything I do has to still work with those.  They also both throw errors during the build when they can't get to the file, but they don't correctly change the path for the browser when they can!

To be clear, none of that is a Sass problem, but a Sass _solution_ would be really nice right about now.

For my scenario, "relative means relative to the file I'm compiling" is correct for both "where it is" and "where it will be," assuming `src\styles.scss` is the sole input file.

**Comment by Roy-Orbison** [comment]
@jneuhaus20

> Regarding
> 
> > Relative URLs in compiled CSS output are resolved relative to the URL the CSS itself is served from, which is something the Sass compiler can't intrinsically know.

It's not a requirement for Sass to know where the final result will be served from in order for us to get an improved experience from Sass, that evidently you also want. Bear in mind that this issue was created as issues like https://github.com/sass/libsass/issues/532 were swept under the rug.

> > I really disagree with this appraisal of the situation. It's like saying unless we have a panacea that covers all edge cases, we can't do basic relative URL resolution.
> 
> I disagree with @Roy-Orbison, bc what @nex3 said is absolutely true and sounds like it was meant to narrow the solution space to what the Sass compiler can reliably do.

What's to disagree with about that statement? Do you honestly believe the Sass compiler can't keep track of a stack of paths even though it already does when resolving `@use` directives? It's not a stretch to (optionally) ask Sass to prefix relative URLs in an include with the paths from that stack. It won't because it's been determined by the devs that the only way to do this is with 'custom URL loaders' (the panacea) and that's clearly not a priority feature, therefore we simply don't get relative resolution. And you don't want plugins to decide, nor does anyone who only wants Sass' output to contain its inputs.

Given this simple tree:
```
src/
├─ assets/
│  ├─ a/
│  │  ├─ background.svg
│  │  ├─ module.scss
│  ├─ b/
│  │  ├─ background.svg
│  │  ├─ module.scss
│  ├─ style.css
│  ├─ style.scss
├─ index.html
```
The logical URL for the SVG in each module file would be just `url(background.svg);`, because then they don't have to know anything about where they'll be included, nor where the assets ends up in the published site resources. Yet Sass will naively have both images referenced in its output as the same `url(background.svg);`, rather than `url(a/background.svg);` and `url(b/background.svg);`. **No one** would actually want URLs that don't resolve to anything, so they're forced to modify both `module.scss` files to contain the path prefix that the filesystem already has (double-specification). It's just absurd that every user would have to use a loader, that would still need to know those `a/` and `b/` prefixes, to get `url(a/background.svg);` in the compiled output. Of course, we can't opt into either feature, because they don't exist. Apparently Sass devs think this data loss is a minor issue, even though Less devs have known for nearly 12 years that it's crucial.

**Comment by nex3** [comment]
> Apparently Sass devs think this data loss is a minor issue, even though Less devs have known for nearly 12 years that it's crucial.

@Roy-Orbison Rein in your tone, or I will ban you from posting in this repository. You are not paying us for our labor, nor are you contributing labor of your own, so you have absolutely no entitlement to get the specific features you want precisely when you want them. We prioritize issues as best we can given extremely limited resources, and unless you want to sponsor development you will wait politely and engage in discussion civilly if you engage in it at all.

**Comment by Roy-Orbison** [comment]
If you feel the need. I'm not waiting for nor expecting anything. I'm just amazed that you still think this is a feature, not a bug. The quantity of issues out there on the 'Net, directly related to this and the closed issues, in all kinds of projects that utilise Sass, suggest otherwise.
